memory_util.h

```diff
@@ -0,0 +1,298 @@
+#pragma once
+
+#include <cassert>
+#include <cstring>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+namespace simulator {
+namespace util {
+
+typedef std::size_t size_type;
+typedef unsigned char uchar;
+
+class BinaryBuffer {
+public:
+    ~BinaryBuffer();
+
+    explicit BinaryBuffer(size_type size = 0);
+    BinaryBuffer(size_type size, size_type capacity);
+    BinaryBuffer(const void* data, size_type size);
+    BinaryBuffer(const BinaryBuffer&);
+
+    BinaryBuffer& operator=(const BinaryBuffer&);
+
+    void assign(const void* data, size_type size);
+
+    void take_over(void* data, size_type size);
+
+    void append(const BinaryBuffer&);
+
+    template <typename T>
+    void append(const T t);
+
+    void append(const std::string& str);
+
+    template <typename T>
+    void append(const std::vector<T>& vec);
+
+    template <typename T>
+    void append(const T* data, int num_elements);
+
+    void start_reading();
+
+    template<typename T>
+    void read(T& t);
+
+    void read(std::string& str);
+
+    template <typename T>
+    void read(std::vector<T>& vec);
+
+    template<typename T>
+    void read(T* t, int num_elements);
+
+    size_type size () const;
+
+    bool resize (size_type);
+
+    size_type capacity () const;
+
+    bool reserve (size_type);
+
+    size_type offset() const;
+
+    bool empty () const;
+
+    bool eof() const;
+
+    void clear ();
+
+    uchar* data_mutable ();
+
+    const uchar* data () const;
+
+private:
+    void delete_if_own() {
+        if (own_) {
+            delete [] data_;
+        }
+    }
+
+    uchar* data_;
+    size_type size_;
+    size_type capacity_;
+    bool own_;
+    uchar* read_ptr_;
+};
+
+inline BinaryBuffer::~BinaryBuffer() {
+    delete_if_own();
+}
+
+inline BinaryBuffer::BinaryBuffer(size_type size) {
+    data_ = (size != 0 ? new uchar[size] : NULL);
+    size_ = capacity_ = size;
+    own_ = true;
+}
+
+inline BinaryBuffer::BinaryBuffer(size_type size, size_type capacity) {
+    if (size > capacity) {
```

Comment:
What's the difference and relationship between size and capacity? Adding a brief explanation to each parameter would be better.